src/dparse/parser.d

```diff
@@ -5301,6 +5326,54 @@ class Parser
         return node;
     }
 
+    /**
+     * Parses attributes of C-style variadic parameters.
+     *
+     * $(GRAMMAR $(RULEDEF variadicArgumentsAttributes):
+     *       $(RULE variadicArgumentsAttribute)+
+     *     ;)
+     */
+    ParameterAttribute[] parseVariadicArgumentsAttributes()
+    {
+        mixin(traceEnterAndExit!(__FUNCTION__));
+        StackBuffer attributes;
+
+        while (moreTokens() && !currentIs(tok!"..."))
+        {
+            if (!attributes.put(parseVariadicArgumentsAttribute()))
+                return null;
+        }
+
+        ParameterAttribute[] buffer;
+        ownArray(buffer, attributes);
+        return buffer;
+    }
+
+    /**
+     * Parses an attribute of C-style variadic parameters.
+     *
+     * $(GRAMMAR $(RULEDEF variadicArgumentsAttribute):
+     *       $(RULE variadicArgumentsAttribute)+
+     *     ;)
+     */
+    ParameterAttribute parseVariadicArgumentsAttribute()
+    {
+        mixin(traceEnterAndExit!(__FUNCTION__));
+        auto node = allocator.make!ParameterAttribute();
+        auto startIndex = index;
+
+        if (!currentIsOneOf(tok!"const", tok!"immutable", tok!"shared", tok!"scope", tok!"return"))
+        {
+            error("`const`, `immutable`, `shared`, `scope`, `return` expected");
+            return null;
+        }
+
+        node.idType = current.type;
+        advance();
+        node.tokens = tokens[startIndex .. index];
```

Comment:
Can't this array only ever have a single element in it? There's only a single call to `advance` here and there aren't any loops in this function.

Reply:
`.tokens` comes from the base AST class which is a range to the tokens some AST node is constructed from, so it's always a range of tokens (and this is the boilerplate similar to all other functions I assume)

Reply:
Tried to follow the usual style, could also write it this way if preferred.

```suggestion
        node.tokens = tokens[index - 1 .. index];
```